src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,273 @@
+        return r;
+    }
+
+    if (!modlist) {
+        SOL_ERR("Failed to find module '%s'", mod_name);
+        return -ENOENT;
+    }
+
+    kmod_list_foreach(itr, modlist) {
+        struct kmod_module *mod;
+        int state, err;
+
+        mod = kmod_module_get_module(itr);
+        state = kmod_module_get_initstate(mod);
+
+        switch (state) {
+        case KMOD_MODULE_BUILTIN:
+            SOL_INF("Module '%s' is builtin", kmod_module_get_name(mod));
+            break;
+
+        case KMOD_MODULE_LIVE:
+            SOL_DBG("Module '%s' is already loaded", kmod_module_get_name(mod));
+            break;
+
+        default:
+            err = kmod_module_probe_insert_module(mod, probe_flags,
+                NULL, NULL, NULL, NULL);
+
+            if (err == 0)
+                SOL_INF("Inserted module '%s'", kmod_module_get_name(mod));
+            else if (err == KMOD_PROBE_APPLY_BLACKLIST)
+                SOL_INF("Module '%s' is blacklisted", kmod_module_get_name(mod));
+            else {
+                SOL_ERR("Failed to insert '%s': %m", kmod_module_get_name(mod));
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static char *
+remove_leading_whitespace(char *input)
+{
+    while (isspace(*input))
+        input++;
+    return input;
+}
+
+static void
+remove_trailing_whitespace(char *input)
+{
+    char *ptr = input + strlen(input) - 1;
+
+    while (ptr > input && isspace(*ptr)) {
+        *ptr = '\0';
+        ptr--;
+    }
+}
+
+static int
+kmod_apply_file(struct kmod_ctx *ctx, int fd)
+{
+    FILE *conf;
+    char buf[1024];
+    int err = 0, line_no = 1;
+
+    conf = fdopen(fd, "re");
+    if (!conf) {
+        close(fd);
+        return -errno;
+    }
+
+    while (fgets(buf, sizeof(buf), conf)) {
+        char *mod_name;
+        int ret;
+
+        mod_name = remove_leading_whitespace(buf);
+        if (*mod_name == '#' || *mod_name == ';' || *mod_name == '\0')
+            continue;
+
+        remove_trailing_whitespace(mod_name);
+        ret = kmod_apply_value(ctx, mod_name);
+        if (ret < 0)
+            err = ret;
+
+        line_no++;
+    }
+
+    fclose(conf);
+    return err;
+}
+
+static int
+kmod_apply_filename(struct kmod_ctx *ctx, int dfd, const char *file_name)
+{
+    int fd;
+
+    if (*file_name == '/' || dfd < 0)
+        fd = open(file_name, O_RDONLY | O_CLOEXEC);
+    else
+        fd = openat(dfd, file_name, O_RDONLY | O_CLOEXEC);
+
+    if (fd < 0) {
+        SOL_DBG("Could not open: %s", file_name);
+        return 0;
+    }
+
+    return kmod_apply_file(ctx, fd);
+}
+
+static int
+kmod_conf_cmp(const void *data1, const void *data2)
+{
+    return strcmp((const char *)data1, (const char *)data2);
+}
+
+static void
+read_directory_to_vector(DIR *dir, struct sol_ptr_vector *vec)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r;
+
+    r = readdir_r(dir, &ent, &entbuf);
```

Comment:
[**@edersondisouza**](https://github.com/edersondisouza) commented on [Oct 5, 2015, 10:04 AM GMT-3](https://github.com/solettaproject/soletta/pull/844#discussion_r41140575 "2015-10-05T13:04:28Z - Replied by Github Comment Enhancer"):
> This `ent` parameter must be allocated in a special fashion, according to `readdir_r(3)`.

The note is more for portable programs. This is Linux-specific, so it should be OK; unless alternative libcs do this differently than glibc, in which case this would need to be checked (specially MUSL, since that's the one we're supporting for `linux-micro` targets).